lib/charms/opensearch/v0/opensearch_locking.py

```diff
@@ -25,78 +25,240 @@
 logger = logging.getLogger(__name__)
 
 
-class OpenSearchOpsLock:
-    """This class covers the configuration changes depending on certain actions."""
-
-    LOCK_INDEX = ".ops_lock"
-    PEER_DATA_LOCK_FLAG = "ops_removing_unit"
+class _PeerRelationEndpoint(ops.Object):
+    _NAME = "node-lock-fallback"
 
-    def __init__(self, charm):
+    def __init__(self, charm: ops.CharmBase):
+        super().__init__(charm, self._NAME)
         self._charm = charm
-        self._opensearch = charm.opensearch
+        self.framework.observe(
+            self._charm.on[self._NAME].relation_changed, self._on_peer_relation_changed
+        )
+
+    @property
+    def acquired(self) -> bool:
+        """Attempt to acquire lock.
 
-    @retry(stop=stop_after_attempt(3), wait=wait_fixed(0.5), reraise=True)
-    def acquire(self):
-        """Method for Acquiring the "ops" lock."""
-        # no lock acquisition needed if only 1 unit remaining
-        if len(self._charm.model.get_relation(PeerRelationName).units) == 1:
+        Returns:
+            Whether lock was acquired
+        """
+        if not self._relation:
```

Comment:
I think this condition never really happens, but if it does, then we should return True instead, no? It means you are alone in the cluster. Maybe we can then check `planned_units > 1`.